File: copyList.java
// TC:SC= O(n)
class Solution {
    HashMap<Node, Node>map;
    public Node copyRandomList(Node head) {
    //null case
        if(head==null) return null;
        
        map= new HashMap();
        Node curr= head;
        Node copyHead= clone(head);
        Node copyCurr= copyHead;
        
        while(curr!= null){             //Single pass- vishay hard!
            copyCurr.next= clone(curr.next);
            copyCurr.random= clone(curr.random);
            curr=curr.next;
            copyCurr=copyCurr.next;
        }
        
        return copyHead;
    }
    
    private Node clone(Node node){
        if (node==null) return null;
        if(map.containsKey(node)) return map.get(node);
        Node newNode= new Node(node.val);
        map.put(node, newNode);
        return map.get(node);
    }
}

File: CopyList.py

# // Time Complexity :  O(N)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this :no

"""

# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""



class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head == None: return head
        curr=head
        currCopy=None

        while(curr!=None):                              
            currCopy = Node(curr.val)
            currCopy.next = curr.next
            curr.next=currCopy
            curr = currCopy.next
        curr=head

        while(curr!=None):                         
            currCopy = curr.next
            if (curr.random):
                currCopy.random = curr.random.next
            curr = currCopy.next
        curr=head

        copynode = curr.next
        currCopy=curr.next
        while(curr.next!=None):                     
            if(currCopy.next==None): break
            curr.next=curr.next.next
            currCopy.next = currCopy.next.next
            curr=curr.next
            currCopy=currCopy.next

        return copynode